webapp/src/app/shared/3d/BabylonViewer.ts

```diff
@@ -0,0 +1,214 @@
+		// light1.specular = new Color3(1, 1, 1).scale(0);
+		
+		performance.mark('start');
+		
+		let layers = map.layers.filter(layer => layer.details.type.toLowerCase() === 'collision');
+		layers.sort((a, b) => a.details.level - b.details.level);
+		
+		// add another layer to the bottom to make the ground visible
+		layers = [await this.generateGroundLayer(layers[0]), ...layers];
+		// layers = [layers[2]];
+		
+		const meshGenerator = new LayerMeshGenerator();
+		
+		const allMeshes: Mesh[] = [];
+		
+		for (let i = 0; i < layers.length; i++) {
+			performance.mark('layerStart');
+			const coll = layers[i];
+			const above = layers[i + 1];
+			let renderAll = 0;
+			if (i === layers.length - 2) {
+				renderAll = 9999;
+			}
+			const layerMaterial = this.textureGenerator.generate(coll.details.level + 1 + renderAll, scene);
+			performance.mark('texture');
+			const meshes = meshGenerator.generateLevel(coll, above, scene);
+			
+			for (const mesh of meshes) {
+				mesh.material = layerMaterial;
+			}
+			
+			allMeshes.push(...meshes);
+			performance.mark('layerEnd');
+			performance.measure('layer: ' + coll.details.level, 'layerStart', 'layerEnd');
+			performance.measure('texture: ' + coll.details.level, 'layerStart', 'texture');
+			performance.measure('mesh: ' + coll.details.level, 'texture', 'layerEnd');
+			
+			
+		}
+		performance.mark('layersEnd');
+		
+		
+		const entityManager = new EntityManager3d();
+		this.entityManager = entityManager;
+		
+		const entityGenerator = new EntityGenerator(entityManager);
+		
+		const entities = map.entityManager.entities;
+		const promises: Promise<any>[] = [];
+		for (const e of entities) {
+			promises.push(entityGenerator.generateEntity(e, scene));
+		}
+		
+		await Promise.all(promises);
+		performance.mark('end');
+		
+		entityManager.init(entityGenerator, scene);
+		
+		const toggle = new ToggleMesh(scene);
+		
+		addWireframeButton(toggle, allMeshes);
+		
+		showAxis(2, scene);
+		
+		performance.measure('layers', 'start', 'layersEnd');
+		performance.measure('entities', 'layersEnd', 'end');
+		
+		this.globalEvents.currentView.next(EditorView.Entities);
+		
+		engine.runRenderLoop(() => scene.render());
+	}
+	
+	
+	private async generateGroundLayer(other: CCMapLayer) {
+		const data: number[][] = [];
+		const height = other.details.height + 10;
+		for (let y = 0; y < height; y++) {
+			data[y] = [];
+			for (let x = 0; x < other.details.width; x++) {
+				data[y][x] = 2;
+			}
+		}
+		const layer = new CCMapLayer(other.getPhaserLayer()!.tilemap);
+		await layer.init({
+			type: 'Collision',
+			name: 'groundColl',
+			level: -1,
+			width: other.details.width,
+			height: height,
+			visible: 1,
+			tilesetName: '',
+			repeat: false,
+			distance: 1,
+			tilesize: Globals.TILE_SIZE,
+			moveSpeed: {x: 0, y: 0},
+			data: data,
+		});
+		
+		for (let y = 0; y < height; y++) {
+			for (let x = 0; x < other.details.width; x++) {
+				layer.getPhaserLayer()!.putTileAt(2, x, y);
+			}
+		}
```

Comment:
This should have already been done in `CCMapLayer.init`. If not, `putTilesAt` should be used.`